File: py/RNG31Transform/Vose.py
from collections import deque

from RNG31Core.AbstractRNG31Core import AbstractRNG31Core
from RNG31Transform.RNGUniform import RNGUniform


class RNGVose:
    """
    Class for returning weighted discrete values

    Technical Explanation: http://web.eecs.utk.edu/~vose/Publications/random.pdf
    Practical Explanation: http://www.keithschwarz.com/darts-dice-coins/

    >>> from GBFlip import GBFlip
    >>> gbFlip = GBFlip(-314159)
    >>> for _ in range(134):
    ...     _ = gbFlip.next()
    >>> vrng = RNGVose(gbFlip, 3, 6, 1, 1, 1)
    >>> vrng.next()
    1
    >>> vrng.next()
    3
    """

    def __init__(self, rng: AbstractRNG31Core, *probabilities):
        assert(rng is not None and isinstance(rng, AbstractRNG31Core))

        # Normalize
        # - sum(probabilities) is always 1.0
        # - each probability is in [0.0, 1.0]
        total = float(sum(probabilities))
        probabilities = [(p / total) for p in probabilities]

        self.__n = len(probabilities)
        avg = 1.0 / self.__n
        small = deque([])
        large = deque([])

        self.__alias = [0.0] * self.__n
        self.__prob = [0.0] * self.__n
        self.__rng = RNGUniform(rng)

        for index in range(self.__n):
            if probabilities[index] >= avg:
                large.append(index)
            else:
                small.append(index)

        while small and large:
            less = small.popleft()
            more = large.popleft()

            self.__prob[less] = probabilities[less] * self.__n
            self.__alias[less] = more

            probabilities[more] = (probabilities[more] +
                                   probabilities[less]) - avg

            if probabilities[more] >= avg:
                large.append(more)
            else:
                small.append(more)

        while small:
            self.__prob[small.popleft()] = 1.0
        while large:
            self.__prob[large.popleft()] = 1.0

    def next(self):
        col = self.__rng.next(self.__n)
        if self.__rng.nextFloat() < self.__prob[col]:
            result = col
        else:
            result = self.__alias[col]
        return result

File: cpp/RNG31Transform/Vose.cpp
#include "Vose.h"

using namespace RNG31;

Vose::Vose(AbstractRNGCore *rng, const double *itemWeights, int itemCount) :
    m_urng(rng)
{
    m_itemCount = itemCount;
    m_probabilities = new double[itemCount];
    m_aliases = new int[itemCount];

    double itemProbabilities[itemCount];
    double sum = 0.0;
    for(int index = 0; index < m_itemCount; ++index)
        sum += itemWeights[index];

    for(int index = 0; index < m_itemCount; ++index)
        itemProbabilities[index] = (m_itemCount * itemWeights[index] / sum);

    std::queue<int> small;
    std::queue<int> large;
    for(int index = 0; index < m_itemCount; ++index)
        if(itemProbabilities[index] < 1.0)
            small.emplace(index);
        else
            large.emplace(index);

    for(uint32_t index = 0; index < small.size(); ++index) {
        small.emplace(small.front());
        small.pop();
    }
    for(uint32_t index = 0; index < large.size(); ++index) {
        large.emplace(large.front());
        large.pop();
    }

    while((!small.empty()) && (!large.empty())) {
        int less = small.front(); small.pop();
        int more = large.front(); large.pop();

        m_probabilities[less] = itemProbabilities[less];
        m_aliases[less] = more;

        // Compute remainder of large
        itemProbabilities[more] += itemProbabilities[less] - 1.0;
        if(itemProbabilities[more] < 1.0)
            small.emplace(more);
        else
            large.emplace(more);
    }

    for(uint32_t index = 0; index < small.size(); ++index) {
        small.emplace(small.front());
        small.pop();
    }
    for(uint32_t index = 0; index < large.size(); ++index) {
        large.emplace(large.front());
        large.pop();
    }

    while(!large.empty()) {
        m_probabilities[large.front()] = 1.0;
        large.pop();
    }
    while(!small.empty())
    {
        m_probabilities[small.front()] = 1.0;
        small.pop();
    }
}

Vose::~Vose()
{
    delete[] m_probabilities;
    m_probabilities = nullptr;

    delete[] m_aliases;
    m_aliases = nullptr;

    m_itemCount = 0;
}

int Vose::next()
{
    int index = m_urng.next(m_itemCount);
    if(m_urng.nextFloat() >= m_probabilities[index])
        index = m_aliases[index];
    return index;
}
